packages/hardhat-core/src/builtin-tasks/compile.ts

```diff
@@ -1016,78 +1005,186 @@ subtask(TASK_COMPILE_SOLIDITY_GET_COMPILATION_JOBS_FAILURE_REASONS)
+          } else if (incompatibleImports.length > 2) {
+            const otherImportsCount = incompatibleImports.length - 2;
+            importsText = ` imports ${incompatibleImports[0]}, ${
+              incompatibleImports[1]
+            } and ${otherImportsCount} other ${pluralize(
+              otherImportsCount,
+              "file"
+            )}. Use --verbose to see the full list.`;
+          }
+
+          errorMessage += `* ${sourceName} (${versionsRange})${importsText}\n`;
+        }
 
-${importsIncompatibleFile.map((x) => `* ${x}`).join("\n")}
+        errorMessage += "\n";
+      }
+
+      if (indirectlyImportsIncompatibleFile.length > 0) {
+        errorMessage += `These files depend on other files that use a different and incompatible version of Solidity:
 
 `;
+
+        for (const error of indirectlyImportsIncompatibleFile) {
+          const { sourceName } = error.file;
+          const { versionPragmas } = error.file.content;
+          const versionsRange = versionPragmas.join(" ");
+
+          const incompatibleImportsFiles: ResolvedFile[] =
+            error.extra?.incompatibleIndirectImports ?? [];
+
+          const incompatibleImports = incompatibleImportsFiles.map(
+            (x: ResolvedFile) =>
+              `${x.sourceName} (${x.content.versionPragmas.join(" ")})`
+          );
+
+          log(
+            `File ${sourceName} depends on files ${incompatibleImportsFiles
+              .map((x) => x.sourceName)
+              .join(", ")} that use an incompatible version of Solidity`
+          );
+
+          let importsText = "";
+          if (incompatibleImports.length === 1) {
+            importsText = ` depends on ${incompatibleImports[0]}`;
+          } else if (incompatibleImports.length === 2) {
+            importsText = ` depends on ${incompatibleImports[0]} and ${incompatibleImports[1]}`;
+          } else if (incompatibleImports.length > 2) {
+            const otherImportsCount = incompatibleImports.length - 2;
+            importsText = ` depends on ${incompatibleImports[0]}, ${
+              incompatibleImports[1]
+            } and ${otherImportsCount} other ${pluralize(
+              otherImportsCount,
+              "file"
+            )}. Use --verbose to see the full list.`;
+          }
+
+          errorMessage += `* ${sourceName} (${versionsRange})${importsText}\n`;
+        }
+
+        errorMessage += "\n";
       }
+
       if (other.length > 0) {
-        reasons += `These files and its dependencies cannot be compiled with your config:
+        errorMessage += `These files and its dependencies cannot be compiled with your config:
 
-${other.map((x) => `* ${x}`).join("\n")}
+${other.map((x) => `* ${x.file.sourceName}`).join("\n")}
 
 `;
       }
 
-      reasons += `Learn more about compiler configuration at https://hardhat.org/config
+      errorMessage += `Learn more about compiler configuration at https://hardhat.org/config
```

Comment:
```suggestion
      errorMessage += `This can be because they have incompatible Solidity pragmas, or don't match any of your configured Solidity compilers. Learn more about compiler configuration at https://hardhat.org/config
```

Reply:
Couldn't that message be misleading? If there are no matching compilers or if there are incompatible imports, we should have detected it and reported it as a different error reason.

Reply:
I believe this error has two reasons:

* The range described by the concatenation of all the pragmas is empty.
* The range described by the concatenation of all the pragmas doesn't include any of the configured compilers, even though each individual pragma does match at least one of the compilers. 

Is that right?

If so, we should aim to explain that without too many details.